Milling/Nuclear Sciences revamp + Lots of minor fixes

draknyte1 opened the pull request:
Added recipes for Isamill and Flotation Cell content.
Added missing Pine Tree textures.
Corrected tooltip on IsaMill controller.
Fixed bad casing allocation.

draknyte1 committed:
Added missing Pine Tree textures.
Added recipes for Isamill and Flotation Cell content.
Corrected tooltip on IsaMill controller.
Fixed bad casing allocation.

draknyte1 committed:
Added new NEI handler for LFTR.
Fixed LFTR fuel recipes.
Fixed LFTR recipe handling.

draknyte1 committed:
Adjusted LFTR tooltip.
Adjusted LFTR sparging behaviour.
Adjusted LFTR U233 output rate.

draknyte1 committed:
Adjusted output timings.

draknyte1 committed:
FFPP now only requires LuV Muffler.
Adjusted LFTR sparge rates, again.
Adjusted costs of Nuclear Salt production.

draknyte1 committed:
Fix a heap of conflicting dehydrator recipes.

draknyte1 committed:
Fixed Tiny dust recipes.

draknyte1 committed:
Fixed shaped dust crafting recipes, turns out they don't like multithreading.

draknyte1 committed:
Fixed NEI Integer overflows.

draknyte1 committed:
Version bump.

draknyte1 committed:
Added recipes for Bronze & Aluminium Machine Casings.
Hid all 8k/32k flasks.

draknyte1 committed:
Added NEI handler for LFTR Sparging.
Fixed Material generation for noble gases.
Fixed Chemical Formula not being applied to some cells.

draknyte1 committed:
Updated NEI sparging handler.

draknyte1 committed:
Rewrote sparge handling to use recipe map.

draknyte1 committed:
Moved all Nuclear recipes to their own class.
Fixed Air Intake working when disabled.

draknyte1 committed:
Greatly improved ICO recipe importation from Pyro Oven.

draknyte1 committed:
Added new generic Nuclear Chem item.
Added ability to set Fluid for a Material.
Moved Nuclear recipe generation to a different phase during init.

draknyte1 committed:
Add workaround for BW Hydrogen Peroxide unification.
Made generic GT++ Recipes maps use default GT NEI handler.
Boosted fuel value of Rocket Fuels.

draknyte1 committed:
Update how Material RGB is applied to GT++ Tools.

draknyte1 committed:
Added more Nuclear fluids/dusts.
Added two new machines.

draknyte1 committed:
Greatly improved NEI handling.
Fixed broken Manure Slurry recipes.

draknyte1 committed:
Did some localizations.

draknyte1 committed:
Better handling of Material chemical formulae.
Fixed some broken recipes.

draknyte1 committed:
Added more nuclear processing recipes.
Removed lots of useless nuclear generated items.
More NEI improvements.
Made Dehydrator and FFPP recipe handling more lenient.

GlodBlock commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_IsaMill.java:
```diff
@@ -344,7 +319,7 @@ public Block getIntakeBlock() {
 
 	public byte getIntakeMeta() {
-		return 0;
+		return 1;
```
it is 0 here in old code. the isamill pipe casing even can't change its texture when it is around the controller.

draknyte1 replied:
And? They're visible from the side and there's a good chance I just never finished the structure check.. why else would I have added them?

draknyte1 committed:
Stopped fluids from having their generation queued.
Possibly improved GT++ Fluid generation where GT counterparts exist.
Adding remapping event for when existing items are removed with a suitable replacement added.
Did some locale work.

draknyte1 committed:
Removed some useless materials.
Fixed Plasma generation and implemented plasma recipes for them.

draknyte1 committed:
More minor fixes.

draknyte1 committed:
Added more LFTR fuel processing.

draknyte1 committed:
More Nuclear Fuel recipe work.

draknyte1 committed:
Added a new multiblock for sparging.
More Nuclear reprocessing work.
Updated Tooltips on all multiblocks.

draknyte1 committed:
Added ability to clear all output hatches on GT++ multis with a plunger.
Changed fluid textures for many autogenerated fluids.
Renamed & remapped Cryolite to Cryolite (F).
Fixed toArray() in AutoMap.java.
Fixed Sparging.
Final touches to Nuclear Fuel reprocessing.

draknyte1 committed:
Updated lots of tooltips.
Made low tier tanks portable, now that higher tier tanks are.
Fixed ULV gens not working correctly.
Disabled ClassTransformer_GT_ItemMachines_Tooltip.

draknyte1 committed:
More Tooltips.
Dehydrator improvements.
Fixed Circuit Programmer not working for Breakthrough Circuits.

draknyte1 committed:
Migrate some recipe from GTNH CoreMod.

draknyte1 committed:
Added HS-188A.
Fixed NEI handling for some handlers.
Fixed some tooltips.

draknyte1 committed:
Fixed hot ingot recipes.

draknyte1 committed:
Fixed ChemPlant book.

draknyte1 committed:
Adjusted LFTR a bit.

Glease commented on src/Java/gregtech/api/util/GasSpargingRecipeMap.java:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	public static GasSpargingRecipe findRecipe(FluidStack aSpargeGas, FluidStack aSpentFuel) {
+		for (GasSpargingRecipe aRecipe : mRecipes) {
```
Preferably use a `Map<FluidStack, GasSpargingRecipe>` to speed up lookup.

draknyte1 replied:
Wouldn’t work, of the three recipes two have the same sparge gas. Didn’t feel it was worth optimising beyond this as I probably won’t ever add more recipes to it either.

Glease replied:
If that's the case, we can just leave this here

Glease commented on src/Java/gtPlusPlus/core/item/bauble/BatteryPackBaseBauble.java:
```diff
@@ -95,7 +95,7 @@ public void addInformation(final ItemStack stack, final EntityPlayer aPlayer, fi
 
 		list.add(EnumChatFormatting.GREEN + aString1 + EnumChatFormatting.GRAY);
-		list.add(EnumChatFormatting.GREEN + aString2+" " + (int) getTransferLimit(stack) + aEUT +" "+ aString3 + EnumChatFormatting.GRAY);
+		list.add(EnumChatFormatting.GREEN + aString2+" " + (int) getTransferLimit(stack) + aEUT +" 7"+ aString3 + EnumChatFormatting.GRAY);
```
?

draknyte1 replied:
Err, I have no idea. 🤣

Glease commented on src/Java/gtPlusPlus/nei/GT_NEI_LFTR_Sparging.java:
```diff
@@ -0,0 +1,377 @@
+		FluidStack tFluid = GT_Utility.getFluidForFilledItem(aResult, true);
+		if (tFluid != null) {
+			tResults.add(GT_Utility.getFluidDisplayStack(tFluid, false));
+			for (FluidContainerRegistry.FluidContainerData tData : FluidContainerRegistry.getRegisteredFluidContainerData()) {
```
This sounds horribly slow. Maybe you can copy over some code from the updated GT_NEI_DefaultHandler.

Glease commented on src/Java/gtPlusPlus/nei/GT_NEI_LFTR_Sparging.java:
```diff
@@ -0,0 +1,377 @@
+		FluidStack tFluid = GT_Utility.getFluidForFilledItem(aInput, true);
+		if (tFluid != null) {
+			tInputs.add(GT_Utility.getFluidDisplayStack(tFluid, false));
+			for (FluidContainerRegistry.FluidContainerData tData : FluidContainerRegistry.getRegisteredFluidContainerData()) {
```
Same here.

Glease reviewed (commented):
I'm still slowly going over these 225 files changed. Currently ~40% completion. I'll left the rest of them to maybe tomorrow

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/basic/GregtechMetaTileEntity_ReactorColdTrap.java:
```diff
@@ -0,0 +1,67 @@
+		super(aID, aName, aNameRegional, aTier, 1, 
+				"Just like the Arctic", 2, 9, "Dehydrator.png", "",
+				new ITexture[]{
+						new GT_RenderedTexture(TexturesGtBlock.OVERLAY_REACTOR_COLDTRAP_SIDE_ACTIVE),
```
This is throwing a bunch of deprecation warnings. I looked a bit and we can get rid of that by changing this to `TextureFactory.of(xxxxxxx)`. Can also remove the explicit array creation.

draknyte1 replied:
I'd like to port 98% of this PR back to my master branch, I have no idea how to use TextureFactory and don't really see a reason to change to it > the existing implementation.

Glease replied:
It has support for glowing (lights will have constant lightness regardless of the environment) and rotation (structurelib), but if you don't care about these and want to merge back to upstream I guess using the old GT_RenderedTexture is ok.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/basic/GregtechMetaTileEntity_ReactorProcessingUnit.java:
```diff
@@ -0,0 +1,67 @@
+		super(aID, aName, aNameRegional, aTier, 1, 
+				"Processes Nuclear things", 2, 9, "Dehydrator.png", "",
+				new ITexture[]{
+						new GT_RenderedTexture(TexturesGtBlock.OVERLAY_REACTOR_PROCESSINGUNIT_SIDE_ACTIVE),
```
Same deprecation warnings here.

draknyte1 committed:
Fixed issue of Multiblock hatches/buses not being assigned recipe maps.
Removed 7 from Charge Pack tooltips.
Initial work on Elemental Duplicator.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/loaders/RecipeGen_DustGeneration.java:
```diff
@@ -38,71 +38,72 @@ public RecipeGen_DustGeneration(final Material M, final boolean O){
-
-
-		if (ItemUtils.checkForInvalidItems(tinyDust) && ItemUtils.checkForInvalidItems(normalDust)) {
+		final ItemStack normalDust = M.getDust(1);
```
Just curious: why do you move these to constructor?

draknyte1 replied:
Because if I didn't they didn't work when the runnable was later called, no idea why.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/recipes/GregtechRecipeAdder.java:
```diff
@@ -1731,8 +1758,77 @@ public boolean addFuelForRTG(ItemStack aFuelPellet, int aFuelDays, int aVoltage)
+				0);		
 
+		int aSize = GT_Recipe_Map.sVacuumRecipes.mRecipeList.size();
+		GT_Recipe_Map.sVacuumRecipes.add(aRecipe);
```
You are not checking for collisions here - can that be problematic?

draknyte1 replied:
It's only used for my hot ingots.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/HANDLER_GT.java:
```diff
@@ -55,6 +55,7 @@
 import gtPlusPlus.xmod.gregtech.loaders.ProcessingToolHeadChoocher;
 import gtPlusPlus.xmod.gregtech.loaders.misc.AddCustomMachineToPA;
 import gtPlusPlus.xmod.gregtech.loaders.recipe.RecipeLoader_AlgaeFarm;
+import gtPlusPlus.xmod.gregtech.loaders.recipe.RecipeLoader_Nuclear;
```
Nothing but a new import?

draknyte1 replied:
I moved things here then out again, fixed next push.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java:
```diff
@@ -342,155 +356,121 @@ public boolean turnCasingActive(final boolean status) {
 		return true;
 	}
 
+	public FluidStack[] getStoredFluidsAsArray() {	
```
Could have just used `getStoredFluids().toArray(new FluidStack[0])`

draknyte1 replied:
Fixed in next push.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java:
```diff
@@ -342,155 +356,121 @@ public boolean turnCasingActive(final boolean status) {
+				}
+			}
+			// Find li2bef4, Helium & Fluorine
 			for (final FluidStack hatchFluid1 : tFluids) { //Loops through hatches
```
We probably don't need to check for existence of Li2BeF4. It's always part of these recipes so findRecipe has done it already.

draknyte1 replied:
Fixed in next push.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java:
```diff
@@ -544,87 +524,139 @@ public void explodeMultiblock() {
+				this.mSpargeTime = 0;
+				this.mSpargeTicks = 0;
+				//Logger.WARNING("Sparging!");
+				//trySparge();
```
I suppose you need to uncomment this?

draknyte1 replied:
Nope, it was originally done in the multi prior to implementing the entire reprocessing chain.

draknyte1 replied:
Fixed in next push.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java:
```diff
@@ -544,87 +524,139 @@ public void explodeMultiblock() {
+		this.depleteInput(depletionStack);
+		updateSlots();		
+		for (int i=0;i<aSpargeRecipe.mFluidOutputs.length;i++) {
+			if (aSpargeRecipe.mInputGas.isFluidEqual(spargeGas)) {
```
What are you trying to do with this if?

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java:
```diff
@@ -544,87 +524,139 @@ public void explodeMultiblock() {
 			}
+			int aGasAmount = MathUtils.randInt(0, aSpargeRecipe.mMaxOutputQuantity[i]);
+			FluidStack aOutput = aSpargeRecipe.mFluidOutputs[i];
+			aSpargeGasAmount -= aGasAmount;
```
Can this ever go negative? Probably need to stop if all input gas is used up.

draknyte1 replied:
Nope, but this isn't used anymore anyway.
I'll just remove it.

draknyte1 replied:
Fixed in next push.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java:
```diff
@@ -544,87 +524,139 @@ public void explodeMultiblock() {
 		else {
-			this.turnCasingActive(false);
-		}
+			// Try output some Uranium-233
```
We should move this else block to onRunningTick for obvious reasons.

Glease replied:
Also, it needs to check if the machine is running with an actual recipe instead of just warming up.

draknyte1 replied:
No, the texture change is done when it's warmed up, regardless of anything else. That's what it symbolizes.

Glease replied:
I mean, this will make it output some URANIUM233 even when the machine is turned off. This is onPostTick.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java:
```diff
@@ -342,155 +356,121 @@ public boolean turnCasingActive(final boolean status) {
-		final Collection<GT_Recipe> tRecipeList = GTPP_Recipe.GTPP_Recipe_Map.sLiquidFluorineThoriumReactorRecipes.mRecipeList;
-		if((tFluids.size() > 0) && (tRecipeList != null)) { //Does input hatch have a LFTR fuel?
+		// Warm up for 4~ minutes
+		Logger.WARNING("Checking LFTR recipes.");
```
This makes it warm up even with no input. Not sure if this is what you want.

draknyte1 replied:
It's fine.

Glease reviewed (commented):
Sparge handling code looks problematic. Will look into it deeper tomorrow.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java:
```diff
@@ -0,0 +1,443 @@
+	@Override
+	public GT_Recipe.GT_Recipe_Map getRecipeMap() {
+		GTPP_Recipe_Map.sSpargeTowerRecipes.mRecipeList.clear();
+		if (GTPP_Recipe_Map.sSpargeTowerRecipes.mRecipeList.isEmpty()) {
```
Clear it last cycle and check if it's empty the next cycle?

draknyte1 replied:
Oh, that was debugging x]

draknyte1 committed:
Removed an unused import from HANDLER_GT.
Balanced FFPP fuel production recipes.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java:
```diff
@@ -0,0 +1,443 @@
+			if (tRecipe != null) {
+				Logger.INFO("Found recipe!");
+				try {
+					String[] aRecipeInfo = RecipeUtils.getRecipeInfo(tRecipe);
```
Preferably not go over this array or generate it if debug logging is turned off.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java:
```diff
@@ -0,0 +1,443 @@
+					this.mOutputFluids = new FluidStack[]{};
+					this.mLastRecipe = tRecipe;		
+					
+					// Deplete Inputs
```
I thought you have depleted the fluids once at L203 by passing true to isRecipeInputEqual?

draknyte1 committed:
Cleaned up GregtechMTE_NuclearReactor.java.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java:
```diff
@@ -0,0 +1,443 @@
+		else {
+			Logger.INFO("Found sparge recipe!");
+			try {
+				String[] aRecipeInfo = aSpargeRecipe.getRecipeInfo();
```
Preferably not go over this array or generate it if debug logging is turned off.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java:
```diff
@@ -0,0 +1,443 @@
+		if (aSpargeGasAmount > 0) {
+			aOutputGases.add(new FluidStack(aSpargeRecipe.mInputGas.getFluid(), aSpargeGasAmount));	
+		}
+		Logger.INFO("Sparge Outputs: "+ItemUtils.getArrayStackNames(aTempMap));
```
Preferably not generate this string if debug logging is turned off.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java:
```diff
@@ -0,0 +1,443 @@
+					this.mMaxProgresstime = Math.max(1, aDevProgress);
+					this.mOutputItems = new ItemStack[]{};
+					ArrayList<FluidStack> aFluidOutputs = getByproductsOfSparge(tRecipe.mFluidInputs[0], tRecipe.mFluidInputs[1]);					
+					this.mOutputFluids = (FluidStack[]) aFluidOutputs.toArray(new FluidStack[0]).clone();
```
No need to clone. I guess you can even make getByproductsOfSparge return an array instead, so we can skip one more array copy here.

Glease reviewed (commented):
I somehow missed this class when I go over the changeset.

draknyte1 committed:
More Cleanup.

draknyte1 committed:
Cleanup.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java:
```diff
@@ -65,21 +63,21 @@
-					ofHatchAdder(GregtechMetaTileEntity_SpargeTower::addOutputToMachineList, getCasingIndex(), 1),
-					ofHatchAdder(GregtechMetaTileEntity_SpargeTower::addInputToMachineList, getCasingIndex(), 1),
-					ofHatchAdder(GregtechMetaTileEntity_SpargeTower::addMaintenanceToMachineList, getCasingIndex(), 1),
+					ofHatchAdder(GregtechMetaTileEntity_SpargeTower::addToMachineList, getCasingIndex(), 1),
```
Duplicates?

draknyte1 replied:
Whatcha mean?

Glease replied:
You are chaining 4 same structure checker.

draknyte1 replied:
Why have four calls when you can just have one? AddToMachineList does the same thing.

draknyte1 replied:
I probably don’t understand structurelib well enough tbh.

Glease replied:
Basically, all these elements passed to ofChain will be tried one by one until one returned true. This is functionally equivalent to `if (this.addToMachineList(getCasingIndex(), 1)))` to what you'd have in old style checkMachine. You can safely drop the other 3 duplicates without breaking anything.

Glease commented on src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_ElementalDuplicator.java:
```diff
@@ -1,6 +1,9 @@
-/*
```
I guess this is still WIP?

draknyte1 replied:
Yeah, Just poked at it a bit earlier to kill time.

draknyte1 committed:
Fix Sir-Mixalot crash.

draknyte1 committed:
Fixed cells being consumed in nuclear reprocessing.
Removed some debug logging.
Removed ClassTransformer_CC_GuiContainerManager.
Improved handling in GT_NEI_LFTR_Sparging.
Added Erbium.
Added HeLiCoPtEr.

draknyte1 committed:
Revert hatch handling in GregtechMetaTileEntity_SpargeTower.

draknyte1 committed:
Add recipes for Sparging Multiblock.
Add some localization.
Migrate material recipes out of item classes.

draknyte1 committed:
Migrate Assembly Line recipes to new OreDict handler.
Added missing file RecipeGen_MetalRecipe.java.

draknyte1 committed:
Add one final milling recipe for Monazite.

draknyte1 committed:
Locale Fixes.

draknyte1 committed:
Merge branch 'master' of https://github.com/GTNewHorizons/GTplusplus into gtnh-milling

# Conflicts:
#	build.properties
#	src/Java/gtPlusPlus/xmod/gregtech/api/metatileentity/implementations/base/generators/GregtechRocketFuelGeneratorBase.java
#	src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/generators/GT_MetaTileEntity_SemiFluidGenerator.java
#	src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/generators/GregtechMetaTileEntityGeothermalGenerator.java
#	src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMTE_NuclearReactor.java

draknyte1 committed:
Post-Conflict Merge fixes.

draknyte1 committed:
Fix minor oversight.

draknyte1 committed:
Revert time adjustment.

draknyte1 committed:
Added recipes for the Cold Trap and Reactor Processing Unit.
Fixed NEI handler for sparging.
Changed ABS NEI handler.
Replace non-verbose FMLCommonHandler.instance().exitJava() calls with CORE.Crash(string).

draknyte1 committed:
Texture change for Sparge Tower based on popular opinion.

draknyte1 committed:
Fixed bug in energy/dynamo handling.

draknyte1 committed:
Fixed GregtechMeta_MultiBlockBase not handling hatches correctly.

draknyte1 committed:
Fix cell being eaten.
